src/command/archive/push/file.c

```diff
@@ -63,47 +64,80 @@ archivePushFile(
-            // If the wal segment already exists in the repo then compare checksums
-            walSegmentFile = walSegmentFind(storageRepo(), archiveId, archiveFile, 0);
-
-            if (walSegmentFile != NULL)
+            // Check each repo for the WAL segment
+            for (unsigned int repoIdx = 0; repoIdx < repoTotal; repoIdx++)
             {
-                String *walSegmentRepoChecksum = strSubN(walSegmentFile, strSize(archiveFile) + 1, HASH_TYPE_SHA1_SIZE_HEX);
+                // If the wal segment already exists in the repo then compare checksums
+                const String *walSegmentFile = walSegmentFind(storageRepoIdx(repoIdx), repoData[repoIdx].archiveId, archiveFile, 0);
 
-                if (strEq(walSegmentChecksum, walSegmentRepoChecksum))
+                if (walSegmentFile != NULL)
                 {
-                    MEM_CONTEXT_PRIOR_BEGIN()
+                    String *walSegmentRepoChecksum = strSubN(walSegmentFile, strSize(archiveFile) + 1, HASH_TYPE_SHA1_SIZE_HEX);
+
+                    // If the checksums are the same then succeed but warn in case this is a symptom of some other issue
+                    if (strEq(walSegmentChecksum, walSegmentRepoChecksum))
                     {
-                        result = strNewFmt(
-                            "WAL file '%s' already exists in the archive with the same checksum"
-                                "\nHINT: this is valid in some recovery scenarios but may also indicate a problem.",
-                            strZ(archiveFile));
+                        MEM_CONTEXT_PRIOR_BEGIN()
+                        {
+                            // Add LF if there has already been a warning
+                            if (result == NULL)
+                                result = strNew("");
+                            else
+                                strCatZ(result, "\n");
+
+                            // Add warning to the result that will be returned to the main process
+                            strCatFmt(
+                                result,
+                                "WAL file '%s' already exists in the repo%u archive with the same checksum"
+                                    "\nHINT: this is valid in some recovery scenarios but may also indicate a problem.",
+                                strZ(archiveFile), cfgOptionGroupIdxToKey(cfgOptGrpRepo, repoIdx));
+                        }
+                        MEM_CONTEXT_PRIOR_END();
+
+                        // No need to copy to this repo
+                        destinationCopy[repoIdx] = false;
+                    }
+                    // Else error so we don't overwrite the existing segment
+                    else
+                    {
+                        THROW_FMT(
+                            ArchiveDuplicateError, "WAL file '%s' already exists in the repo%u archive", strZ(archiveFile),
+                            cfgOptionGroupIdxToKey(cfgOptGrpRepo, repoIdx));
```

Comment:
Yes -- because if we just go ahead and write the current segment to the other repos then we will have different versions of the segment scattered across the repos. Just as now, if they want to push a new version of the segment they'll need to remove the old one manually. In practice this indicates split brain and must be dealt with.